src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,150 @@
+    }
 
     @Override
     public void add(T value) {
-
+        if (!ensureCapacity()) {
+            Object[] newArray = expand();
+            System.arraycopy(elementData, 0, newArray, 0, size);
+            elementData = newArray;
+        }
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index != size) {
+            indexCheck(index);
+        }
+        if (ensureCapacity() && index <= size) {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+            size++;
+            return;
+        }
+        elementData[size++] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] copyOfList = new Object[list.size()];
+        for (int i = 0; i < copyOfList.length; i++) {
+            copyOfList[i] = list.get(i);
+        }
+        if (ensureCapacity()) {
+            System.arraycopy(copyOfList, 0, elementData, size, copyOfList.length);
+        }
+        Object[] newArray = expand();
+        System.arraycopy(elementData, 0, newArray, 0, size);
+        System.arraycopy(copyOfList, 0, newArray, size, copyOfList.length);
+        elementData = newArray;
+        size += copyOfList.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        T elementRemoved = (T) elementData[index];
+        int moveNumberOfElements = size - index - 1;
+        System.arraycopy(elementData, index + 1, elementData, index, moveNumberOfElements);
+        elementData[--size] = null;
+        return elementRemoved;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T elementRemoved = null;
+        for (int i = 0; i < size; i++) {
+            if (element == null && elementData[i] == null
+                    || element != null && element.equals(elementData[i])) {
+                elementRemoved = remove(i);
+                if (elementRemoved == null) {
+                    return null;
+                }
+            }
+        }
+        if (elementRemoved == null) {
+            throw new NoSuchElementException("Element: " + element + " was not found");
+        }
+        return elementRemoved;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private boolean ensureCapacity() {
+        return this.size < elementData.length;
+    }
+
+    private <T> Object[] expand() {
+        return (T[]) new Object[elementData.length + (elementData.length >> 1)];
```

Comment:
what does `>> 1` mean?

Reply:
right shift operator, moving bits to the right by one position, as a result, we're getting half a number. Same for negative numbers, please correct me if I'm wrong